stream-parent/stream-client/src/main/java/org/interledger/stream/sender/SimpleStreamSender.java

```diff
@@ -964,100 +971,94 @@ void setUnrecoverableErrorEncountered(boolean unrecoverableErrorEncountered) {
-    //       rejectPacket -> handleReject(preparePacket, streamPacket, rejectPacket, prepareAmounts, numRejectedPackets,
-    //           congestionController)
-    //   );
-    //
-    //   // Mark the streamConnection object as closed if the caller supplied a ConnectionCloseFrame
-    //   streamFrames.stream()
-    //       .filter(streamFrame -> streamFrame.streamFrameType() == StreamFrameType.ConnectionClose)
-    //       .findAny()
-    //       .ifPresent($ -> {
-    //         streamConnection.closeConnection();
-    //         logger.info("STREAM Connection closed.");
-    //       });
-    //
-    //   // Emit a log statement if the called supplied a StreamCloseFrame
-    //   streamFrames.stream()
-    //       .filter(streamFrame -> streamFrame.streamFrameType() == StreamFrameType.StreamClose)
-    //       .findAny()
-    //       .map($ -> (StreamCloseFrame) $)
-    //       .ifPresent($ -> {
-    //         logger.info(
-    //             "StreamId {} Closed. Delivered: {} ({} packets fulfilled, {} packets rejected)",
-    //             $.streamId(), paymentTracker.getDeliveredAmountInReceieverUnits(), this.numFulfilledPackets.get(),
-    //             this.numRejectedPackets.get()
-    //         );
-    //       });
-    // }
+//     void closeConnection() throws StreamConnectionClosedException {
+//       this.sendStreamFramesInZeroValuePacket(Lists.newArrayList(
+//           ConnectionCloseFrame.builder()
+//               .errorCode(ErrorCodes.NoError)
+//               .build()
+//       ));
+//     }
+
+     private void sendStreamFramesInZeroValuePacket(final Collection<StreamFrame> streamFrames)
+         throws StreamConnectionClosedException {
+       Objects.requireNonNull(streamFrames);
+
+       if (streamFrames.size() <= 0) {
+         logger.warn("sendStreamFrames called with 0 frames");
+         return;
+       }
+
+       final StreamPacket streamPacket = StreamPacket.builder()
+           .interledgerPacketType(InterledgerPacketType.PREPARE)
+           .prepareAmount(UnsignedLong.ZERO)
+           .sequence(streamConnection.nextSequence())
+           .addAllFrames(streamFrames)
+           .build();
+
+       // Create the ILP Prepare packet using an encrypted StreamPacket as the encryptedStreamPacket payload...
+       final byte[] encryptedStreamPacket = this.toEncrypted(sharedSecret, streamPacket);
+       final InterledgerCondition executionCondition;
+       executionCondition = generatedFulfillableFulfillment(sharedSecret, encryptedStreamPacket).getCondition();
+
+       final InterledgerPreparePacket preparePacket = InterledgerPreparePacket.builder()
+           .destination(destinationAddress)
+           .amount(UnsignedLong.ZERO)
+           .executionCondition(executionCondition)
+           .expiresAt(DateUtils.now().plusSeconds(30L))
+           .data(encryptedStreamPacket)
+           .typedData(streamPacket)
+           .build();
+
+       link.sendPacket(preparePacket);
```

Comment:
Should this not be done in  a handler, like this?

```
link.sendPacket(preparePacket).handle(
                  fulfillPacket -> ...,
                  rejectPacket -> ...
```

Put another way, if the `link.sendPacket` fails, should lines 1042-1061 still occur? If the answer is "yes" then those should be moved above line 1039. If the answer is "no", then this seems fine, but feels more natural to only close the stream if the other side replies that it actually got those frames.

Reply:
Whether they fulfill or reject, our sender is done using the stream connection. It's more like an FYI to the receiver at this point. I didn't see the point in adding any logic there. But maybe I don't understand this well enough to see the reason that's incorrect.

Reply:
Ok, in that case lines 1041-1061 should come before `link.sendPacket` on line 1039. Rationale is that if line 1039 fails, then the `streamConnection` won't actually be marked as closed.

Reply:
I just remove or comment out lines 1041-1061 (again) because my changes only send a StreamClose frame but not the ConnectionClose frame.  I only uncommented the code to send a StreamClose frame. The code to send ConnectionClose frame is still commented out because that felt like it had bigger issues according the the github issue #308. I don't understand the implications, but I would've expected that I shouldn't use the stream connection after it's been closed. And moving lines 1041-1046 before sending the packet would do just that.

Reply:
commenting out for now